### PR TITLE
Interactivity plugin cleanup and optimizations

### PR DESCRIPTION
**Pull request by michealparks**
I've been wanting to clean up the interactivity after modeling the XR repo's pointer controls after it, and I was able to get to it this week!

I've done a few simple cleanup / optimization tasks here:

* Organize contexts and use symbols for keys
* Use a weakmap instead of userData for event dispatching
* Other small cleanup and optimization that I'll list in comments

### CODE REVIEW

File: packages/extras/src/lib/interactivity/context.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+import {
```
Comment (michealparks) [inline packages/extras/src/lib/interactivity/context.ts:1]: Moved all context-related logic here

File: packages/extras/src/lib/interactivity/context.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  setContext<InteractivityContext>(contextKey, context)
+
+  return context
+}
+
+interface HandlerContext {
+  dispatchers: WeakMap<THREE.Object3D, ReturnType<typeof createRawEventDispatcher<ThrelteEvents>>>
+}
+
+const handlerContextKey = Symbol('interactivity-handler-context')
```
Comment (michealparks) [inline packages/extras/src/lib/interactivity/context.ts:84]: Created a new context for dispatchers

File: packages/extras/src/lib/interactivity/defaults.ts [hunk 3]
```diff
@@ -1,8 +1,10 @@
-import { memoize, useThrelte, watch } from '@threlte/core'
-import type { ComputeFunction, State } from './types'
+import { useThrelte, watch, type CurrentWritable } from '@threlte/core'
+import type { ComputeFunction } from './context'
 
-export const getDefaultComputeFunction = (state: State): ComputeFunction => {
-  const camera = memoize(useThrelte().camera)
+export const getDefaultComputeFunction = (
+  target: CurrentWritable<HTMLElement>
```
Comment (michealparks) [inline packages/extras/src/lib/interactivity/defaults.ts:5]: This function takes only the target store to simplify context creation

File: packages/extras/src/lib/interactivity/setupInteractivity.ts [hunk 4]
```diff
@@ -61,60 +58,46 @@ export const setupInteractivity = (state: State) => {
...
-      // Filter out duplicates
-      .filter((item) => {
-        const id = getIntersectionId(item as Intersection)
-        if (duplicates.has(id)) return false
-        duplicates.add(id)
-        return true
-      })
-
-    if (state.filter) hits = state.filter(hits, state)
+    const hits = context.raycaster.intersectObjects(context.interactiveObjects, true)
```
Comment (michealparks) [inline packages/extras/src/lib/interactivity/setupInteractivity.ts:67]: The `intersectObjects` method sorts by distance and contains no duplicates as part of its behavior, allowing us to simplify `getHits()`

Comment (DefinitelyMaybe) [reply packages/extras/src/lib/interactivity/setupInteractivity.ts:67]: [can confirm](https://threejs.org/docs/index.html?q=ray#api/en/core/Raycaster.intersectObjects), [source link too](https://github.com/mrdoob/three.js/blob/841d2e791d3e8a2463322c5ca31b16956828b91c/src/core/Raycaster.js#L66)

File: packages/extras/src/lib/interactivity/setupInteractivity.ts [hunk 5]
```diff
@@ -1,15 +1,10 @@
...
-  return object.userData._threlte_interactivity_dispatcher as
-    | ReturnType<typeof createRawEventDispatcher<ThrelteEvents>>
-    | undefined
-}
+import { getHandlerContext, type InteractivityContext } from './context'
+import type { DomEvent, Intersection, IntersectionEvent } from './types'
 
 function getIntersectionId(event: Intersection) {
-  return (event.eventObject || event.object).uuid + '/' + event.index + event.instanceId
+  return `${(event.eventObject || event.object).uuid}/${event.index}${event.instanceId}`
```
Comment (DefinitelyMaybe) [inline packages/extras/src/lib/interactivity/setupInteractivity.ts:7]: nice! this looks clearer to me as to what's going on here

File: packages/extras/src/lib/interactivity/setupInteractivity.ts [hunk 6]
```diff
@@ -61,60 +58,46 @@ export const setupInteractivity = (state: State) => {
     }
   }
 
-  const enabled = memoize(state.enabled)
+  const enabled = memoize(context.enabled)
 
   const getHits = (): Intersection[] => {
-    const duplicates = new Set<string>()
+    if (!enabled) return []
```
Comment (grischaerbe) [inline packages/extras/src/lib/interactivity/setupInteractivity.ts:64]:
```suggestion
    if (!enabled.current) return []
```

**Review by grischaerbe — CHANGES_REQUESTED** [review]
I have a tiny change, apart from that it looks fabulous!

File: packages/extras/src/lib/interactivity/setupInteractivity.ts (hunk 6 above)
Comment (michealparks) [reply packages/extras/src/lib/interactivity/setupInteractivity.ts:64]: Ah! Good catch. Should have gone with an explicit comparison 🤦

**Pull request merged**